staging/src/k8s.io/kubectl/pkg/cmd/debug/profiles.go

```diff
@@ -49,11 +62,195 @@ func profileLegacy(pod *corev1.Pod, containerName string, target runtime.Object)
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.SecurityContext = nil
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		setHostNamespace(pod, false)
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.SecurityContext = nil
+			c.SecurityContext = addCap(c.SecurityContext, "SYS_PTRACE")
+		})
+
+	case styleEphemeral:
+		for i := range pod.Spec.EphemeralContainers {
+			container := &pod.Spec.EphemeralContainers[i]
+			if container.Name != containerName {
+				continue
+			}
+			// For ephemeral container: empty securityContext
+			container.SecurityContext = nil
+		}
+
+	case styleNode:
+		// empty securityContext; uses isolated namespaces
+		pod.Spec.SecurityContext = nil
+		setHostNamespace(pod, false)
+
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileBaseline, target)
+	}
+
+	return nil
+}
+
+func (p *restrictedProfile) Apply(pod *corev1.Pod, containerName string, target runtime.Object) error {
+	// drops privilege and run as non-root
+	sc := &corev1.SecurityContext{
+		RunAsNonRoot: pointer.BoolPtr(true),
+		Capabilities: &corev1.Capabilities{
+			Drop: []corev1.Capability{"ALL"},
+		},
+	}
+
+	switch getDebugStyle(pod, target) {
+	case stylePodCopy:
+		// For copy of pod: empty securityContext; sets shareProcessNamespace
+		pod.Spec.ShareProcessNamespace = pointer.BoolPtr(true)
+		modifyContainer(pod.Spec.Containers, containerName, func(c *corev1.Container) {
+			c.SecurityContext = sc
+		})
+
+	case styleEphemeral:
+		modifyEphemeralContainer(pod.Spec.EphemeralContainers, containerName, func(c *corev1.EphemeralContainer) {
+			c.SecurityContext = sc
+		})
+
+	case styleNode:
+		// no additional settings required other than the common
+		// settings applied below
+
+	default:
+		return fmt.Errorf("the %s profile doesn't support objects of type %T", ProfileRestricted, target)
+	}
+
+	// common settings, empty securityContext and uses isolated namespaces
+	pod.Spec.SecurityContext = nil
+	setHostNamespace(pod, false)
+
+	return nil
+}
+
+func setHostNamespace(pod *corev1.Pod, enabled bool) {
+	pod.Spec.HostNetwork = enabled
+	pod.Spec.HostPID = enabled
+	pod.Spec.HostIPC = enabled
+}
+
+func addCap(s *corev1.SecurityContext, c corev1.Capability) *corev1.SecurityContext {
+	if s == nil {
+		return &corev1.SecurityContext{
+			Capabilities: &corev1.Capabilities{
+				Add: []corev1.Capability{c},
+			},
+		}
+	}
+
+	ss := s.DeepCopy()
+	if ss.Capabilities == nil {
+		ss.Capabilities = &corev1.Capabilities{
+			Add: []corev1.Capability{c},
+		}
+		return ss
+	}
+
+	ss.Capabilities.Add = append(ss.Capabilities.Add, c)
+	return ss
+}
+
+// modifyContainer performs m against a container from cs which has the name of containerName.
+func modifyContainer(cs []corev1.Container, containerName string, m func(*corev1.Container)) {
```

Comment:
I think it would be better to keep this given that most, if not all, profiles currently make use of this boilerplate code.

The name `modifyContainer` makes it clear at call sites and this does.